Generate unique synthetic queries

besimali opened the pull request:
The issue this addresses (didn't want to link it with original repo):

```
Instead of synthetic generation being on the control nodes on a fixed schedule, the synthetic generation should be dynamic for each query. This means all synthetic queries can be unique.

Synthetic generations will need to be really quick, else we might not be able to keep up. Imagine there are 10 image-to-image requests per second you must send off. We naturally can't get a new image for each one, or do any sort of heavy processing else we wont be able to keep up with demand. So some components might need to be cached, but every query should be unique.

So, instead of pulling a synthetic query from Redis, we could generate it on the query node directly. Something to think about there - what if the dataset needed is a few hundred MB, will this cause issues?
```

# What is done?

- Removed the synthetic generation from the control node
- Removed periodic synthetic query generation
- Replaced fetching the synthetic data from Redis in the query node with on-demand generation
- Each text prompt in a synth query is now unique
- Images are cached in Redis. Now that cache is accessed a lot, disk read might become a bottleneck. It makes sense to keep them in memory. Since the query nodes are going to scale, we keep the cache in Redis so they can all access it
- tested both with unit tests and by running the control and query node with postgres and redis. Lots of things were commented out and fake tasks are made in the control node. Verified that the Redis is used as intended by putting log statements in the get_random_image_b64 func.

# What could be improved
- Markov model is trained and loaded for each query node replica. This is due to me not realizing the query node should scale up. The markov model could be trained only once. We could also move the synthetic generation out of the query node so it can be scaled up/down independently. We could then use redis queues to pass the generated data to the query node 
- profiling the current generation would expose the bottlenecks
- the masks are decoded and encoded each time. We could at least avoid decoding by caching the image size, but also encoding by caching the mask

Commit:
initial change. Tests pass

namoray commented:
Some thoughts about scaling: Does this scale well if everything is on the query node?

- Memory usage is now proportional to # query nodes since the images are in memory
- What if the dataset needed to generate the synthetic queries is big (>=100mb)?

besimali commented:
Thanks for the feedback!
I didnt realize the query nodes should scale on the same machine. My bad.

Since we might have big datasets and we want to sample from those. I am thinking of loading these to Redis and then sampling from there. For images we can have a list on redis and access a random image by index. 
I will try to implement that tomorrow

Commit:
use redis for image cache

besimali commented:
Updated to use Redis so the image cache is not replicated for each query node. 
Also updated the PR description to point out some improvements to this that I don't have time to implement now

namoray commented:
Nice description & I agree with how we could improve.

namoray commented on validator/utils/synthetic/synthetic_generation_funcs.py:
```diff
@@ -1,7 +1,7 @@
 import random
 import sys
-from typing import Any
+from typing import Union
```
use | instead of Union

namoray commented on validator/utils/synthetic/synthetic_generation_funcs.py:
```diff
@@ -280,5 +285,6 @@ async def generate_synthetic_data(task: str) -> Any:
 
     func = getattr(sys.modules[__name__], generative_function_name)
     kwargs = task_config.synthetic_generation_config.kwargs
+    kwargs['redis_db'] = redis_db  # Add redis_db to the kwargs
```
pretty obvious that is what is happening without the comment.

Comment could just explain why if you felt like we need one

namoray commented on validator/utils/synthetic/synthetic_utils.py:
```diff
@@ -109,22 +97,37 @@ async def _get_random_picsum_image(x_dim: int, y_dim: int) -> str:
+    logger.debug(f"Starting get_random_image_b64. Cache key: {cache_key}, Cache size: {cache_size}")
+
+    if random.random() < 0.01:
+        if await redis_db.scard(cache_key) > 0:
```
This could maybe be a function as it's not clear to me what this does without reading line by line and trying to work it out.


if random.random() < 0.01:
 _do_some_logic()
etc

namoray commented on validator/utils/synthetic/synthetic_utils.py:
```diff
@@ -109,22 +97,37 @@ async def _get_random_picsum_image(x_dim: int, y_dim: int) -> str:
+            random_key = await redis_db.srandmember(cache_key)
+            await redis_db.srem(cache_key, random_key)
+            await redis_db.delete(random_key)
+    elif await redis_db.scard(cache_key) > 0:
```
Same this isn't the clearest of conditions, would might a comment here

namoray commented on validator/utils/synthetic/synthetic_utils.py:
```diff
@@ -109,22 +97,37 @@ async def _get_random_picsum_image(x_dim: int, y_dim: int) -> str:
+    # If cache is empty or we decided to get a new image
     random_picsum_image = await _get_random_picsum_image(1024, 1024)
-    cache.add(key=str(uuid.uuid4()), value=random_picsum_image)
+    if await redis_db.scard(cache_key) >= cache_size:
```
I think it's slightly confusing because of the use of 'scard' etc. 'scard' is a bad name from redis to me because it's not clear whats happening. Overall this func is a little hard to read

namoray reviewed (approved):
I think this is a very minimal implementation. Great work!

Seems to cover every base asked for, kept to the code style. Even added a test!

If this works well, then awesome work.

besimray committed:
apply review suggestions
